invenio_records_resources/services/records/queryparser/query.py

```diff
@@ -61,19 +61,55 @@ class SearchOptions:
             )
     """
 
-    def __init__(self, identity=None, extra_params=None, tree_transformer_factory=None):
+    def __init__(self, identity=None, extra_params=None, tree_transformer_cls=None):
         """Initialise the parser."""
         self.identity = identity
+        self.tree_transformer_cls = tree_transformer_cls
         self.extra_params = extra_params or {}
-        self.tree_transformer_factory = tree_transformer_factory
+        # the pop or {} is needed due to extra_params being passed from the factory
+        # it is possible that e.g. allow_list=None and then it will fail to set()
+        self.mapping = self.extra_params.pop("mapping", None) or {}
+        self.allow_list = self._calculate_allow_list(
+            self.extra_params.pop("allow_list", None), self.mapping
+        )
+        # process fields list
+        self.fields = self._calculate_fields(
+            # get instead of pop since fields could desired for
+            # both querystring and multi_match
+            self.extra_params.get("fields", None),
+            self.allow_list,
+        )
+
+    def _calculate_allow_list(self, allow_list, mapping):
+        """Calculate the allow list."""
+        if allow_list:  # only add the mapping if there is an allow list
+            # mapping should be at least a subset of the allow list
+            return set(allow_list).union(mapping.values())
```

Comment:
if I understand correctly, here we prefer calculating the allow list with the mappings, instead of failing if there is a mismatch, right?
I would honestly prefer a runtime failure (assert?), so that you know that they are misconfigured. I fear that, with this implementation, the behaviour will not be predictable, as the final list is computed merging the two.

It will indeed require being explicit and duplicate the list of fields.

So, instead, I would:
`assert mapping.values().issubset(set(allow_list)), raise ... a nice error msg ...`

Reply:
On the one hand I agree, on the other hand it depends if we want to favor explicit or implicit configuration. We have three parameters affecting the allowed list: `fields`, `mapping` and `allowed list`. The reasons why I went implicit are:
- Some times the mapping is very very long, then we would need to duplicate a very long list, which is error prone to missing some. For example [ZenodoRDM](https://github.com/zenodo/zenodo-rdm/blob/master/invenio.cfg#L145).
- In my opinion it does not make sense to define a mapping, if you do not want to allow searching over it. Therefore, I see the allow list as an extra to "remove" the capabilities over fields that are not mentioned. For example: we want to map `user: user.username` and then remove the query over `email` (so we do not add it). It is true that this implementing a blacklist by omision.

Reply:
Discussed IRL, implicit is okay due to long list of mappings.